src/generate_price_reconciliation_report.py

```diff
@@ -0,0 +1,264 @@
+        holdings_government_bonds_df_null_isin.join_asof(
+            bond_prices_df,
+            left_on="DATE",
+            right_on="DATETIME",
+            by=["SEDOL"],
+            strategy="backward",
+        )
+        .with_columns(
+            [(pl.col("FUND_PRICE") - pl.col("REF_PRICE")).alias("PRICE_DIFF")]
+        )
+        .select(
+            [
+                "FUND",
+                "DATE",
+                "FINANCIAL_TYPE",
+                "SYMBOL",
+                "SECURITY_NAME",
+                "ISIN",
+                "SEDOL",
+                "FUND_PRICE",
+                "REF_PRICE",
+                "PRICE_DIFF",
+            ]
+        )
+    )
+    equity_reconciliation = (
+        holdings_equities_df.join_asof(
+            equity_prices_df,
+            left_on="DATE",
+            right_on="DATETIME",
+            by=["SYMBOL"],
+            strategy="backward",
+        )
+        .with_columns(
+            [(pl.col("FUND_PRICE") - pl.col("REF_PRICE")).alias("PRICE_DIFF")]
+        )
+        .select(
+            [
+                "FUND",
+                "DATE",
+                "FINANCIAL_TYPE",
+                "SYMBOL",
+                "SECURITY_NAME",
+                "ISIN",
+                "SEDOL",
+                "FUND_PRICE",
+                "REF_PRICE",
+                "PRICE_DIFF",
+            ]
+        )
+    )
+    reconciliation_report = pl.concat(
+        [
+            bond_reconciliation_null_sedol,
+            bond_reconciliation_null_isin,
+            equity_reconciliation,
+        ]
+    ).sort(["FUND", "DATE", "FINANCIAL_TYPE", "ISIN", "SEDOL", "SYMBOL"])
+    return reconciliation_report
+
+
+def generate_report(
+    database_path: str,
+    output_data_path: str,
+    fund_holdings_table_name: str,
+    bond_prices_table_name: str,
+    bond_reference_table_name: str,
+    equity_prices_table_name: str,
+) -> None:
+    """
+    Generates a price reconciliation report for fund holdings.
+
+    Args:
+        database_path (str): Path to the database.
+        output_data_path (str): Path to the output directory.
+        fund_holdings_table_name (str): Name of the fund holdings table.
+        bond_prices_table_name (str): Name of the bond prices table.
+        bond_reference_table_name (str): Name of the bond reference table.
+        equity_prices_table_name (str): Name of the equity prices table.
+
+    Returns:
+        None
+    """
+    conn = Utils.connect_to_database(database_path)
+    bond_prices_df = get_bond_prices(
+        conn, bond_prices_table_name, bond_reference_table_name
+    )
+    fund_holdings_df = Utils.get_data_from_database(conn, fund_holdings_table_name)
+    equity_prices_df = Utils.get_data_from_database(conn, equity_prices_table_name)
+    report_df = perform_price_reconciliation(
+        fund_holdings_df, bond_prices_df, equity_prices_df
+    )
+    if not os.path.exists(output_data_path):
+        os.makedirs(output_data_path)
```

Comment:
`os.makedirs(output_data_path, exist_ok=True)` would mean you don't need the `if`.